plugins/py3embed/py3embed.cpp

```diff
@@ -0,0 +1,1720 @@
+
+// Use class OwnedPyObject for any objects that are not 'borrowed references'
+// so that the appropriate Py_DECREF call is made when the OwnedPyObject goes
+// out of scope, even if the function returns prematurely (such as via an exception).
+// In particular, checkPythonError is a lot easier to call safely if this is used.
+
+class OwnedPyObject
+{
+    PyObject *ptr;
+public:
+    inline OwnedPyObject() : ptr(NULL)     {}
+    inline OwnedPyObject(PyObject *_ptr) : ptr(_ptr) {}
+    inline ~OwnedPyObject()                { if (ptr) Py_DECREF(ptr); }
+    inline PyObject * get() const           { return ptr; }
+    inline PyObject * getClear()            { PyObject *ret = ptr; ptr = NULL; return ret; }
+    inline PyObject * operator -> () const { return ptr; }
+    inline operator PyObject *() const    { return ptr; }
+    inline void clear()                     { if (ptr) Py_DECREF(ptr); ptr = NULL; }
+    inline void setown(PyObject *_ptr)      { clear(); ptr = _ptr; }
+    inline void set(PyObject *_ptr)         { if (_ptr) Py_INCREF(_ptr); clear(); ptr = _ptr; }
+    inline PyObject *getLink()              { if (ptr) Py_INCREF(ptr); return ptr;}
+    inline PyObject **ref()                 { return &ptr; }
+};
+
+template <class X>
+class OwnedPyX
+{
+    X *ptr;
+public:
+    inline OwnedPyX<X>() : ptr(NULL) {}
+    inline OwnedPyX<X>(X *_ptr) : ptr(_ptr) {}
+    inline ~OwnedPyX<X>()            { if (ptr) Py_DECREF(ptr); }
+    inline X * get() const           { return ptr; }
+    inline X * getClear()            { PyObject *ret = ptr; ptr = NULL; return ret; }
+    inline X * operator -> () const  { return ptr; }
+    inline operator X *() const      { return ptr; }
+    inline void clear()              { if (ptr) Py_DECREF(ptr); ptr = NULL; }
+    inline void setown(X *_ptr)      { clear(); ptr = _ptr; }
+    inline void set(X *_ptr)         { if (_ptr) Py_INCREF(_ptr); clear(); ptr = _ptr; }
+    inline X *getLink()              { if (ptr) Py_INCREF(ptr); return ptr;}
+    inline X **ref()                 { return &ptr; }
+};
+
+
+__declspec(noreturn) static void failx(const char *msg, ...) __attribute__((format(printf, 1, 2), noreturn));
+
+static void failx(const char *message, ...)
+{
+    va_list args;
+    va_start(args,message);
+    StringBuffer msg;
+    msg.append("pyembed: ").valist_appendf(message,args);
+    va_end(args);
+    rtlFail(0, msg.str());
+}
+
+// call checkPythonError to throw an exception if Python error state is set
+
+static void checkPythonError()
+{
+    PyObject* err = PyErr_Occurred();
+    if (err)
+    {
+        OwnedPyObject pType, pValue, pTraceBack;
+        PyErr_Fetch(pType.ref(), pValue.ref(), pTraceBack.ref());
+        OwnedPyObject valStr = PyObject_Str(pValue);
+        PyErr_Clear();
+        assertex(PyUnicode_Check(valStr));
+        OwnedPyObject temp_bytes = PyUnicode_AsEncodedString(valStr, ASCII_LIKE_CODEPAGE, "ignore");
+        failx("%s", PyBytes_AsString(temp_bytes));
```

Comment:
Yes, I wasn't sure whether ascii or utf8 was more appropriate here though in practice I doubt it makes a lot of difference. I'll change to utf8 if you prefer?